vhdbuilder/packer/test/pam/test_pam.py

```diff
@@ -0,0 +1,492 @@
+
+    def expect_login_prompt(self, user):
+        print("\n-----Expecting login prompt")
+        self.expect([r"[Ll]ogin: "], timeout=1)
+        print("\n-----Sending username")
+        self.sendline(user)
+
+    def expect_password_prompt(self, pw):
+        print("\n-----Expecting password prompt")
+        self.expect([r"[Pp]assword: "], timeout=1)
+        print("\n-----Sending password")
+        self.waitnoecho()
+        self.sendline(pw)
+
+    def expect_bash_prompt(self):
+        print("\n-----Expecting bash prompt")
+        self.expect([r"\$ "], timeout=1)
+        print("\n-----Found bash prompt")
+
+
+def check_user_exists(username):
+    """
+    Check if the specified user exists
+
+    Input Arguments:
+      username (str): The name of the user to check
+
+    Returns:
+      True if the user exists, False if not
+    """
+    return any(user[0] == username for user in pwd.getpwall())
+
+
+def gen_pw(length=24, use_lowercase=True, use_uppercase=True, use_numbers=True,
+           use_specials=True):
+    """
+    Generate a random password
+
+    Input Arguments:
+      length (int): The length of the password to generate
+      use_lowercase (bool): Whether to use lowercase characters
+      use_uppercase (bool): Whether to use uppercase characters
+      use_numbers (bool): Whether to use numbers
+      use_specials (bool): Whether to use special characters
+
+    Returns:
+      The generated password
+    """
+    print("\n-----Generating password")
+    charset = ''
+    pw = ''
+
+    # Ensure at least one of each type of character requested character type
+    # is used
+    if use_lowercase:
+        charset += string.ascii_lowercase
+        pw += random.choice(string.ascii_lowercase)
+    if use_uppercase:
+        charset += string.ascii_uppercase
+        pw += random.choice(string.ascii_uppercase)
+    if use_numbers:
+        charset += string.digits
+        pw += random.choice(string.digits)
+    if use_specials:
+        charset += "!@#"
+        pw += random.choice("!@#")
+    if len(pw) > length:
+        raise Exception("Password length is greater than specified length")
+
+    # Add random characters from the character set until the password is the
+    # specified length
+    length -= len(pw)
+    pw += ''.join(random.choices(charset, k=length))
+
+    return pw
+
+
+def set_password(username, password):
+    """
+    Set the password for the specified user
+
+    Input Arguments:
+      username (str): The name of the user to set the password for;
+      password (str): The password to set for the user
+    """
+    print("\n=== Setting password for " + username + " to " + password)
+    child = BashSpawn(f"sudo passwd {username}", encoding='utf-8',
+                      logfile=sys.stdout)
+    try:
+        # child.expect_optional_sudo_auth()
+        child.expect_password_prompt(password)
+        if child.expect(["Retype new password:"]) == 0:
```

Comment:
You can get rid of the `if` and `else` here -- `expect` will raise an exception if it doesn't find a match.